### DISCUSSION TITLE
Does the current API support linearization (fast web view)

### DISCUSSION THREAD

**Discussion by lowell-lewis** [gettalong/hexapdf#226] [Q&A] [answered]
We are evaluating hexapdf and one feature we need is the ability to linearize existing PDFs.

Thanks, Grant

**Comment by gettalong** [comment, marked as answer]
The linearization feature of PDF (annex F in the PDF specification) is not (yet) implemented in HexaPDF. The simple reason for this is it was not requested by someone until now :grin: 

Edit: One thing to add though, implementing linearization is quite a big task for a, nowadays, usually not often used feature.

**Reply by lowell-lewis** [reply]
I won't pretend to be an expert on the PDF specification, our project is investigating what we can do to speed up viewing of large PDF documents. My understanding is a PDF must be linearized if you want to leverage range requests to process a PDF in chunks from a browser. If that is incorrect by all means correct me.

**Reply by gettalong** [reply]
You are correct, this is what linearized PDF was designed for. However, since that time more than 20 years ago internet connections got much faster and even bigger PDFs with sizes in the lower 10s of MiBs won't take that long to fully load.

One tool I know that supports linearization is [qpdf](https://qpdf.sourceforge.io/), maybe using that as a post-processing step would fit your needs?

**Reply by lowell-lewis** [reply]
Thanks, we'll take a look at qpdf. We are also investigating ghostscript to linearize. We have to deal with PDFs that are 100s of meg and government issued equipment which is not the fastest.

**Reply by jrochkind** [reply]
> for a, nowadays, usually not often used feature.

I'm curious to learn more about why linearization might not be often used these days. I am still learning about this, but got the impression that it was important for optimal viewing of PDFs over the web, which is of course still a popular thing to do.

**Reply by gettalong** [reply]
Yeah, this is the primary use case, i.e. viewing PDFs over the web (which is also why it is sometimes found under *fast web view*). And there are certainly many PDFs out there that use that feature. However, since many open source solutions don't implement it, many of the generated PDFs are not using linearization. Even very popular libraries like Apache pdfbox or LibrePDF's OpenPDF (based on iText 4) don't support linearization (as far as I was able to determine).

I guess most purely commercial solutions support linearization and also some open source tools, like qpdf. So postprocessing is always a possibility.
